bob-apps/bin/bobc.rs

```diff
@@ -1,71 +1,365 @@
+            }
+            let v: Vec<u64> = v.into_iter().map(|n| n.parse().unwrap()).collect();
+            let range = v[0]..=v[1];
+            return Ok(KeyPattern::Range(range));
+        }
+        Ok(KeyPattern::Multiple(
+            key_arg_cont
+                .split(",")
+                .map(|n| n.parse().unwrap())
+                .collect(),
+        ))
+    }
+}
+
+#[derive(Debug)]
+enum FilePattern<'a> {
+    WithRE(RePattern<'a>, PathBuf),
+    WithoutRE(&'a str),
+}
+
+#[derive(Debug)]
+struct RePattern<'a> {
+    inner: &'a str,
+}
+
+impl<'a> RePattern<'a> {
+    fn new(f: &'a str) -> Self {
+        Self { inner: f }
+    }
+
+    fn get_regex(&self) -> Regex {
+        lazy_static! {
+            static ref RE: Regex = Regex::new("\\\\\\{key\\\\\\}").unwrap();
+        }
+        let st = RE
+            .replace(&regex::escape(self.inner), "(\\d+)")
+            .into_owned();
+        Regex::new(&st).unwrap()
+    }
+    fn get_filenames(&self, keys: &KeyPattern) -> Vec<(u64, String)> {
+        lazy_static! {
+            static ref RE: Regex = Regex::new("\\{key\\}").unwrap();
+        }
+        match keys {
+            KeyPattern::Single(val) => {
+                vec![(*val, RE.replace(self.inner, val.to_string()).into_owned())]
+            }
+            KeyPattern::Range(r) => {
+                r.clone() // idk how to use map Range without moving out a value
+                    .map(|n| (n, RE.replace(self.inner, n.to_string()).into_owned()))
+                    .collect()
+            }
+            KeyPattern::Multiple(v) => v
+                .into_iter()
+                .map(|n| (*n, RE.replace(self.inner, n.to_string()).into_owned()))
+                .collect(),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum KeyPattern {
+    Single(u64),
+    Range(RangeInclusive<u64>),
+    Multiple(Vec<u64>),
+}
+
+#[derive(Debug)]
+struct KeyName {
+    key: Vec<u8>,
+    name: String,
+}
+
 #[tokio::main]
 async fn main() {
     env_logger::builder().filter_level(LevelFilter::Info).init();
     let matches = get_matches();
-    if let (sc, Some(sub_mathes)) = matches.subcommand() {
-        let key = get_key_value(sub_mathes);
-        match sc {
-            "put" => {
-                let value = sub_mathes.value_of("size").unwrap();
-                let size = value.parse().expect("size must be usize");
-                let addr = sub_mathes
-                    .value_of("uri")
-                    .expect("has default value")
-                    .parse()
-                    .expect("wrong format of url");
-                info!("PUT key: \"{:?}\" size: \"{}\" to \"{}\"", key, size, addr);
-                put(key, size, addr).await;
+    let app_args = AppArgs::from_matches(&matches);
+
+    let addr = NetConfig::from_args(&app_args).get_uri();
+    match app_args.subcommand {
+        PUT_SC => {
+            let keys_names = match app_args.file_pattern {
```

Comment:
Probably the code will be clearer if match on tuple from `file_pattern` and `key_pattern`: 
`match (app_args.file_pattern, app_args.key_pattern) { .... }`

Reply:
On the contrary, I've tried, it was a lot messier.

Reply:
On second thought: [1f3b6fe](https://github.com/qoollo/bob/pull/562/commits/1f3b6fe4575408884c3271f030c38333d1327eee)